src/main/scala/is/hail/annotations/ReadRowsRDD.scala

```diff
@@ -0,0 +1,134 @@
+package is.hail.annotations
+
+import java.io.DataInputStream
+
+import is.hail.HailContext
+import is.hail.expr.TStruct
+import is.hail.utils.{SerializableHadoopConfiguration, _}
+import net.jpountz.lz4.{LZ4BlockInputStream, LZ4Factory}
+import org.apache.commons.lang3.StringUtils
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import org.apache.spark.{Partition, SparkContext, TaskContext}
+
+import scala.annotation.meta.param
+
+class RichRDDRow(val rdd: RDD[Row]) extends AnyVal {
+  def writeRows(path: String, t: TStruct) {
+    val sc = rdd.sparkContext
+    val hadoopConf = sc.hadoopConfiguration
+
+    hadoopConf.mkDir(path + "/rowstore")
+
+    val tBc = sc.broadcast(t)
+    val sHadoopConfBc = sc.broadcast(new SerializableHadoopConfiguration(hadoopConf))
+
+    val nPartitions = rdd.partitions.length
+    val d = digitsNeeded(nPartitions)
+
+    val rowCount = rdd.mapPartitionsWithIndex { case (i, it) =>
+      val region = MemoryBuffer(8 * 1024)
+      val rvb = new RegionValueBuilder(region)
+      val buffer = new Array[Byte](8 * 1024)
+      var rowCount = 0L
+
+      val t = tBc.value
+      val f = t.fundamentalType
+
+      val is = i.toString
+      assert(is.length <= d)
+      val pis = StringUtils.leftPad(is, d, "0")
+
+      sHadoopConfBc.value.value.writeLZ4DataFile(path + "/rowstore/part-" + pis,
+        64 * 1024,
+        LZ4Factory.fastestInstance().highCompressor()) { out =>
+        it.foreach { r =>
+          region.clear()
+          rvb.start(f)
+          rvb.addRow(t, r)
+          val offset = rvb.end()
+          assert(offset == 0)
+
+          val rowSize = region.offset
+          out.writeInt(rowSize)
+
+          var totalWritten = 0
+          while (totalWritten < rowSize) {
+            val toWrite = buffer.length.min(rowSize - totalWritten)
+            region.loadBytes(totalWritten, toWrite, buffer)
+            out.write(buffer, 0, toWrite)
+            totalWritten += toWrite
+          }
+
+          rowCount += 1
+        }
+
+        out.writeInt(-1)
+      }
+
+      Iterator(rowCount)
+    }
+      .fold(0L)(_ + _)
+
+    info(s"wrote $rowCount records")
+  }
+}
+
+case class ReadRowsRDDPartition(index: Int) extends Partition
+
+class ReadRowsRDD(@(transient @param) sc: SparkContext,
```

Comment:
https://stackoverflow.com/questions/3769624/annotating-constructor-parameters-in-scala

Oops, this should actually be `@(transient @field)`.  Will fix.